src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexInterpreter.cs

```diff
@@ -223,6 +224,20 @@ private bool MatchString(string str)
             {
                 if (runtextend - runtextpos < c)
                 {
+                    // If MatchString was called after a greedy op such as a .*, we would have zipped runtextpos to the end without really examining any characters. Reset to maxBacktrackPos here as an optimization
+                    if (_maxBacktrackPosition != -1 && runtextpos > _maxBacktrackPosition)
+                    {
+                        // If lastIndexOf is -1, we backtrack to the max extent possible.
+                        runtextpos = _maxBacktrackPosition;
+                        ReadOnlySpan<char> runtextSpan = runtext.AsSpan(_maxBacktrackPosition);
+                        int lastIndexOf = runtextSpan.LastIndexOf(str);
+                        if (lastIndexOf > -1)
+                        {
+                            // Found the next position to match. Move runtextpos here
+                            runtextpos = _maxBacktrackPosition + lastIndexOf;
```

Comment:
How is this affected if there are nested loops, each of which might be doing backtracking?  This doesn't need to be a stack of positions?

Reply:
Got an example of what you mean by nested loops?

Reply:
Ignore my question... I just realized this is only being done for `Not` and not other kinds of loops that might contain loops.